### PR TITLE
implement physical_limit.cpp for support limit

### PR DESCRIPTION
**Pull request by KKould**
### What problem does this PR solve?

- implement physical_limit.cpp & physical_merge_limit.cpp
- fixed build logical_plan for limit
- add python test script for limit
- support stream sink for physical_sink.cpp

Issue link: https://github.com/infiniflow/infinity/issues/362

### What is changed and how it works?

```
kould=> SELECT * FROM test_limit limit 10;
 c1 | c2 
----+----
  0 |  1
  2 |  3
  4 |  5
  6 |  7
(4 rows)

kould=> SELECT * FROM test_limit limit 2 offset 2;
 c1 | c2 
----+----
  4 |  5
  6 |  7
(2 rows)
```

### Code changes

- [x] Has Code change
- [ ] Has CI related scripts change

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Note for reviewer

### CODE REVIEW

File: src/executor/operator/physical_limit.cppm [hunk 1]
```diff
@@ -29,43 +30,83 @@ export module physical_limit;
...
+
+    // Returns the right index after limit
+    virtual SizeT Limit(SizeT row_count) = 0;
+
+    virtual bool IsLimitOver() = 0;
+};
+
+export class AtomicCounter : public LimitCounter {
+public:
+    AtomicCounter(const i64 &offset, const i64 &limit) : offset_(offset), limit_(limit) {}
```
Comment (JinHai-CN) [inline src/executor/operator/physical_limit.cppm:48]: Don't use reference.

File: src/executor/operator/physical_limit.cppm [hunk 2]
```diff
@@ -29,43 +30,83 @@ export module physical_limit;
...
+    bool IsLimitOver();
+
+private:
+    ai64 offset_;
+    ai64 limit_;
+};
+
+export class UnSyncCounter : public LimitCounter {
+public:
+    UnSyncCounter(const i64 &offset, const i64 &limit) : offset_(offset), limit_(limit) {}
```
Comment (JinHai-CN) [inline src/executor/operator/physical_limit.cppm:63]: Don't use reference to pass i64

File: src/executor/operator/physical_limit.cpp [hunk 3]
```diff
@@ -34,130 +38,201 @@ module physical_limit;
 
 namespace infinity {
 
-void PhysicalLimit::Init() {}
+SizeT AtomicCounter::Offset(SizeT row_count) {
+    auto success = false;
+    SizeT result;
```
Comment (JinHai-CN) [inline src/executor/operator/physical_limit.cpp:43]: It is better to provide a initialized value to all local variables.

File: src/executor/operator/physical_limit.cpp [hunk 4]
```diff
@@ -34,130 +38,201 @@ module physical_limit;
...
+        if (input_blocks[block_id]->row_count() == 0) {
+            continue;
+        }
+        SizeT max_offset = input_blocks[block_id]->row_count() - 1;
 
-    SharedPtr<TableDef> table_def = TableDef::Make(MakeShared<String>("default"), MakeShared<String>("limit"), columns);
-    SharedPtr<DataTable> output_table = DataTable::Make(table_def, TableType::kIntermediate);
+        if (offset > max_offset) {
+            offset -= max_offset;
+            continue;
```
Comment (JinHai-CN) [inline src/executor/operator/physical_limit.cpp:193]: Is this 'continue' redundant?

File: src/executor/operator/physical_merge_limit.cppm [hunk 5]
```diff
@@ -29,31 +31,32 @@ namespace infinity {
...
         return 0;
     }
 
 private:
-    SharedPtr<Vector<String>> output_names_{};
-    SharedPtr<Vector<SharedPtr<DataType>>> output_types_{};
+    SharedPtr<BaseExpression> limit_expr_{};
+    SharedPtr<BaseExpression> offset_expr_{};
+
+    SharedPtr<LimitCounter> counter_;
```
Comment (JinHai-CN) [inline src/executor/operator/physical_merge_limit.cppm:59]: How about UniquePtr ?

File: src/executor/operator_state.cpp [hunk 6]
```diff
@@ -99,6 +102,18 @@ bool QueueSourceState::GetData() {
...
             break;
         }
+        case PhysicalOperatorType::kMergeLimit: {
+            auto *fragment_data = static_cast<FragmentData *>(fragment_data_base.get());
+            MergeLimitOperatorState *limit_op_state = (MergeLimitOperatorState *)next_op_state;
+            limit_op_state->input_data_blocks_.push_back(Move(fragment_data->data_block_));
+            if (!limit_op_state->input_complete_) {
+                limit_op_state->input_complete_ = completed;
+            }
+            if (limit_op_state->input_complete_) {
```
Comment (JinHai-CN) [inline src/executor/operator_state.cpp:112]: Merge this if to above one?

File: src/executor/operator/physical_limit.cpp [hunk 7]
```diff
@@ -34,130 +38,201 @@ module physical_limit;
...
-                total_row_count -= block_row_count;
-            } else {
-                start_block = block_id;
-                start_row_id = total_row_count;
-                rest_row_count = block_row_count - total_row_count;
-                break;
-            }
-        }
+SizeT PhysicalLimit::TaskletCount() {
+    i64 limit = (static_pointer_cast<ValueExpression>(limit_expr_))->GetValue().value_.big_int;
```
Comment (KKould) [inline src/executor/operator/physical_limit.cpp:159]: wrong count

File: src/executor/operator_state.cppm [hunk 8]
```diff
@@ -407,31 +417,34 @@ export enum class SinkStateType {
...
 
     [[nodiscard]] inline SinkStateType state_type() const { return state_type_; }
 
+    [[nodiscard]] inline bool IsMaterialize() const { return fragment_type_ == FragmentType::kSerialMaterialize || fragment_type_ == FragmentType::kParallelMaterialize; }
+
     inline bool Error() const { return error_message_.get() != nullptr; }
 
     u64 fragment_id_{};
     u64 task_id_{};
+    FragmentType fragment_type_{};
```
Comment (JinHai-CN) [inline src/executor/operator_state.cppm:433]: How about pass fragment_context into PhysicalSink::Execute method?

File: src/executor/operator/physical_merge_limit.cppm (hunk 5 above)
Comment (KKould) [reply src/executor/operator/physical_merge_limit.cppm:59]: If you use `AtomicCounter`, SharedPtr is required to ensure sharing

File: src/executor/operator/physical_limit.cppm [hunk 9]
```diff
@@ -29,43 +30,85 @@ export module physical_limit;
...
 
-    static SharedPtr<DataTable> GetLimitOutput(const SharedPtr<DataTable> &input_table, i64 limit, i64 offset);
-
 private:
     SharedPtr<BaseExpression> limit_expr_{};
     SharedPtr<BaseExpression> offset_expr_{};
 
-    SharedPtr<DataTable> input_table_{};
-    u64 input_table_index_{};
+    UniquePtr<LimitCounter> counter_;
```
Comment (JinHai-CN) [inline src/executor/operator/physical_limit.cppm:111]: Init it with nullptr

File: src/executor/operator/physical_merge_limit.cppm [hunk 10]
```diff
@@ -29,31 +31,32 @@ namespace infinity {
...
         return 0;
     }
 
 private:
-    SharedPtr<Vector<String>> output_names_{};
-    SharedPtr<Vector<SharedPtr<DataType>>> output_types_{};
+    SharedPtr<BaseExpression> limit_expr_{};
+    SharedPtr<BaseExpression> offset_expr_{};
+
+    UniquePtr<LimitCounter> counter_;
```
Comment (JinHai-CN) [inline src/executor/operator/physical_merge_limit.cppm:59]: Init counter_

**Review by JinHai-CN — APPROVED** [review]

**Pull request merged**